x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java

```diff
@@ -1249,60 +1255,137 @@ public String getFeatureDescription() {
         return "Provides anomaly detection and forecasting functionality";
     }
 
-    @Override public void cleanUpFeature(
+    @Override
+    public void cleanUpFeature(
         ClusterService clusterService,
         Client client,
-        ActionListener<ResetFeatureStateResponse.ResetFeatureStateStatus> listener) {
+        ActionListener<ResetFeatureStateResponse.ResetFeatureStateStatus> finalListener) {
+        logger.info("Starting machine learning cleanup");
+
+        ActionListener<ResetFeatureStateResponse.ResetFeatureStateStatus> unsetResetModeListener = ActionListener.wrap(
+            success -> client.execute(SetResetModeAction.INSTANCE, SetResetModeAction.Request.disabled(), ActionListener.wrap(
+                resetSuccess -> finalListener.onResponse(success),
+                resetFailure -> {
+                    logger.warn("failed to disable reset mode after state clean up success", resetFailure);
+                    finalListener.onResponse(success);
+                })
+            ),
+            failure -> client.execute(SetResetModeAction.INSTANCE, SetResetModeAction.Request.disabled(), ActionListener.wrap(
+                resetSuccess -> finalListener.onFailure(failure),
+                resetFailure -> {
+                    logger.warn("failed to disable reset mode after state clean up failure", resetFailure);
+                    finalListener.onFailure(failure);
+                })
+            )
+        );
 
         Map<String, Boolean> results = new ConcurrentHashMap<>();
 
+        ActionListener<ListTasksResponse> afterWaitingForTasks = ActionListener.wrap(
+            listTasksResponse -> {
+                listTasksResponse.rethrowFailures("Waiting for indexing requests for .ml-* indices");
```

Comment:
If this throws `unsetResetModeListener` will not be called is that intentional? All other paths call `unsetResetModeListener`

Reply:
https://github.com/elastic/elasticsearch/blob/318ae89e14b9b1b925d33e139177a8e51297d12c/server/src/main/java/org/elasticsearch/action/ActionListener.java#L128-L150

Is the definition of the action listener created. The contents are all wrapped in a try catch